{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "CTCG_RegressionExample_Siamese.ipynb",
      "version": "0.3.2",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "jDF-OsZ9HGHr",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "from __future__ import division, print_function"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "05c7-4X4HvcA",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "81c9bc20-bf15-418e-943f-a75c13c25f39"
      },
      "source": [
        "#download the human genome\n",
        "#Get hg38 fasta by download 2bit and then converting to fa\n",
        "#On the cluster, this file is stored in /mnt/data/annotations/...\n",
        "![[ -f hg38.2bit ]] || wget http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.2bit -O hg38.2bit  \n",
        "![[ -f twoBitToFa ]] || wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/twoBitToFa -O twoBitToFa\n",
        "!chmod a+x twoBitToFa\n",
        "![[ -f hg38.genome.fa ]] || ./twoBitToFa hg38.2bit hg38.genome.fa\n",
        "\n",
        "#download hg38 chromsizes file\n",
        "# This would also be in /mnt/data/annotations\n",
        "![[ -f hg38.chrom.sizes ]] || wget http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes -O hg38.chrom.sizes\n",
        "  \n",
        "#install bedtools\n",
        "!apt-get install bedtools\n",
        "\n",
        "#Install samtools\n",
        "%cd /content\n",
        "![[ -f samtools-1.9.tar.bz2 ]] || wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2\n",
        "!tar -xjf samtools-1.9.tar.bz2\n",
        "%cd samtools-1.9\n",
        "!./configure\n",
        "!make\n",
        "!make install\n",
        "%cd ..\n",
        "\n",
        "#make index for the human genome\n",
        "#These are already built for the files on the cluster\n",
        "![[ -e hg38.genome.fa.fai ]] || samtools faidx hg38.genome.fa"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-07-15 19:44:04--  http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.2bit\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 835393456 (797M) [text/plain]\n",
            "Saving to: ‘hg38.2bit’\n",
            "\n",
            "hg38.2bit           100%[===================>] 796.69M  68.9MB/s    in 15s     \n",
            "\n",
            "2019-07-15 19:44:19 (53.9 MB/s) - ‘hg38.2bit’ saved [835393456/835393456]\n",
            "\n",
            "--2019-07-15 19:44:20--  http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/twoBitToFa\n",
            "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 5886992 (5.6M) [text/plain]\n",
            "Saving to: ‘twoBitToFa’\n",
            "\n",
            "twoBitToFa          100%[===================>]   5.61M  2.67MB/s    in 2.1s    \n",
            "\n",
            "2019-07-15 19:44:22 (2.67 MB/s) - ‘twoBitToFa’ saved [5886992/5886992]\n",
            "\n",
            "--2019-07-15 19:44:37--  http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 11672 (11K) [text/plain]\n",
            "Saving to: ‘hg38.chrom.sizes’\n",
            "\n",
            "hg38.chrom.sizes    100%[===================>]  11.40K  --.-KB/s    in 0s      \n",
            "\n",
            "2019-07-15 19:44:37 (155 MB/s) - ‘hg38.chrom.sizes’ saved [11672/11672]\n",
            "\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-410\n",
            "Use 'apt autoremove' to remove it.\n",
            "The following NEW packages will be installed:\n",
            "  bedtools\n",
            "0 upgraded, 1 newly installed, 0 to remove and 4 not upgraded.\n",
            "Need to get 577 kB of archives.\n",
            "After this operation, 2,040 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 bedtools amd64 2.26.0+dfsg-5 [577 kB]\n",
            "Fetched 577 kB in 0s (1,650 kB/s)\n",
            "Selecting previously unselected package bedtools.\n",
            "(Reading database ... 130963 files and directories currently installed.)\n",
            "Preparing to unpack .../bedtools_2.26.0+dfsg-5_amd64.deb ...\n",
            "Unpacking bedtools (2.26.0+dfsg-5) ...\n",
            "Setting up bedtools (2.26.0+dfsg-5) ...\n",
            "/content\n",
            "--2019-07-15 19:44:42--  https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2\n",
            "Resolving github.com (github.com)... 140.82.114.3\n",
            "Connecting to github.com (github.com)|140.82.114.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/3666841/fe586164-8a73-11e8-84ad-bb90bbd3b7c0?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190715%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190715T194442Z&X-Amz-Expires=300&X-Amz-Signature=d59108f12d3e0a31a0d027b027b6f14ce09a2187fa1b36ae0e94fe4c643fb0f6&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dsamtools-1.9.tar.bz2&response-content-type=application%2Foctet-stream [following]\n",
            "--2019-07-15 19:44:42--  https://github-production-release-asset-2e65be.s3.amazonaws.com/3666841/fe586164-8a73-11e8-84ad-bb90bbd3b7c0?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190715%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190715T194442Z&X-Amz-Expires=300&X-Amz-Signature=d59108f12d3e0a31a0d027b027b6f14ce09a2187fa1b36ae0e94fe4c643fb0f6&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dsamtools-1.9.tar.bz2&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.204.155\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.204.155|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4440405 (4.2M) [application/octet-stream]\n",
            "Saving to: ‘samtools-1.9.tar.bz2’\n",
            "\n",
            "samtools-1.9.tar.bz 100%[===================>]   4.23M  19.5MB/s    in 0.2s    \n",
            "\n",
            "2019-07-15 19:44:43 (19.5 MB/s) - ‘samtools-1.9.tar.bz2’ saved [4440405/4440405]\n",
            "\n",
            "/content/samtools-1.9\n",
            "checking for gcc... gcc\n",
            "checking whether the C compiler works... yes\n",
            "checking for C compiler default output file name... a.out\n",
            "checking for suffix of executables... \n",
            "checking whether we are cross compiling... no\n",
            "checking for suffix of object files... o\n",
            "checking whether we are using the GNU C compiler... yes\n",
            "checking whether gcc accepts -g... yes\n",
            "checking for gcc option to accept ISO C89... none needed\n",
            "checking for special C compiler options needed for large files... no\n",
            "checking for _FILE_OFFSET_BITS value needed for large files... no\n",
            "checking location of HTSlib source tree... htslib-1.9\n",
            "checking for NcursesW wide-character library... no\n",
            "checking for Ncurses library... yes\n",
            "checking for working ncurses/curses.h... no\n",
            "checking for working ncurses.h... yes\n",
            "checking for zlib.h... yes\n",
            "checking for inflate in -lz... yes\n",
            "checking for library containing regcomp... none required\n",
            "configure: creating ./config.status\n",
            "config.status: creating config.mk\n",
            "config.status: creating config.h\n",
            "=== configuring in htslib-1.9 (/content/samtools-1.9/htslib-1.9)\n",
            "configure: running /bin/bash ./configure --disable-option-checking '--prefix=/usr/local'  --cache-file=/dev/null --srcdir=.\n",
            "checking for gcc... gcc\n",
            "checking whether the C compiler works... yes\n",
            "checking for C compiler default output file name... a.out\n",
            "checking for suffix of executables... \n",
            "checking whether we are cross compiling... no\n",
            "checking for suffix of object files... o\n",
            "checking whether we are using the GNU C compiler... yes\n",
            "checking whether gcc accepts -g... yes\n",
            "checking for gcc option to accept ISO C89... none needed\n",
            "checking for ranlib... ranlib\n",
            "checking for grep that handles long lines and -e... /bin/grep\n",
            "checking for C compiler warning flags... -Wall\n",
            "checking for special C compiler options needed for large files... no\n",
            "checking for _FILE_OFFSET_BITS value needed for large files... no\n",
            "checking for _LARGEFILE_SOURCE value needed for large files... no\n",
            "checking shared library type for unknown-Linux... plain .so\n",
            "checking how to run the C preprocessor... gcc -E\n",
            "checking for egrep... /bin/grep -E\n",
            "checking for ANSI C header files... yes\n",
            "checking for sys/types.h... yes\n",
            "checking for sys/stat.h... yes\n",
            "checking for stdlib.h... yes\n",
            "checking for string.h... yes\n",
            "checking for memory.h... yes\n",
            "checking for strings.h... yes\n",
            "checking for inttypes.h... yes\n",
            "checking for stdint.h... yes\n",
            "checking for unistd.h... yes\n",
            "checking for stdlib.h... (cached) yes\n",
            "checking for unistd.h... (cached) yes\n",
            "checking for sys/param.h... yes\n",
            "checking for getpagesize... yes\n",
            "checking for working mmap... yes\n",
            "checking for gmtime_r... yes\n",
            "checking for fsync... yes\n",
            "checking for drand48... yes\n",
            "checking whether fdatasync is declared... yes\n",
            "checking for fdatasync... yes\n",
            "checking for library containing log... -lm\n",
            "checking for zlib.h... yes\n",
            "checking for inflate in -lz... yes\n",
            "checking for library containing recv... none required\n",
            "checking for bzlib.h... yes\n",
            "checking for BZ2_bzBuffToBuffCompress in -lbz2... yes\n",
            "checking for lzma.h... yes\n",
            "checking for lzma_easy_buffer_encode in -llzma... yes\n",
            "checking for libdeflate.h... no\n",
            "checking for libdeflate_deflate_compress in -ldeflate... no\n",
            "checking for curl_easy_pause in -lcurl... yes\n",
            "checking for CCHmac... no\n",
            "checking for library containing HMAC... -lcrypto\n",
            "checking whether PTHREAD_MUTEX_RECURSIVE is declared... yes\n",
            "configure: creating ./config.status\n",
            "config.status: creating config.mk\n",
            "config.status: creating htslib.pc.tmp\n",
            "config.status: creating config.h\n",
            "config.mk:45: htslib-1.9/htslib_static.mk: No such file or directory\n",
            "cd htslib-1.9 && make htslib_static.mk\n",
            "make[1]: Entering directory '/content/samtools-1.9/htslib-1.9'\n",
            "sed -n '/^static_libs=/s/[^=]*=/HTSLIB_static_LIBS = /p;/^static_ldflags=/s/[^=]*=/HTSLIB_static_LDFLAGS = /p' htslib.pc.tmp > htslib_static.mk\n",
            "make[1]: Leaving directory '/content/samtools-1.9/htslib-1.9'\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_index.o bam_index.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_plcmd.o bam_plcmd.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam_view.o sam_view.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_cat.o bam_cat.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_md.o bam_md.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_reheader.o bam_reheader.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_sort.o bam_sort.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bedidx.o bedidx.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_rmdup.o bam_rmdup.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_rmdupse.o bam_rmdupse.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_mate.o bam_mate.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_stat.o bam_stat.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_color.o bam_color.c\n",
            "echo '#define SAMTOOLS_VERSION \"1.9\"' > version.h\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bamtk.o bamtk.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam2bcf.o bam2bcf.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam2bcf_indel.o bam2bcf_indel.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sample.o sample.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o cut_target.o cut_target.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o phase.o phase.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam2depth.o bam2depth.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o padding.o padding.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bedcov.o bedcov.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bamshuf.o bamshuf.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o faidx.o faidx.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o dict.o dict.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o stats.o stats.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o stats_isize.o stats_isize.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_flags.o bam_flags.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_split.o bam_split.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_tview.o bam_tview.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_tview_curses.o bam_tview_curses.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_tview_html.o bam_tview_html.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_lpileup.o bam_lpileup.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_quickcheck.o bam_quickcheck.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_addrprg.o bam_addrprg.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_markdup.o bam_markdup.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o tmp_file.o tmp_file.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o lz4/lz4.o lz4/lz4.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_aux.o bam_aux.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam.o bam.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_import.o bam_import.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam.o sam.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam_header.o sam_header.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o bam_plbuf.o bam_plbuf.c\n",
            "ar -csru libbam.a bam_aux.o bam.o bam_import.o sam.o sam_header.o bam_plbuf.o\n",
            "ar: `u' modifier ignored since `D' is the default (see `U')\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam_opts.o sam_opts.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o sam_utils.o sam_utils.c\n",
            "ar -rcs libst.a sam_opts.o sam_utils.o\n",
            "cd htslib-1.9 && make lib-static\n",
            "make[1]: Entering directory '/content/samtools-1.9/htslib-1.9'\n",
            "gcc -Wall -g -O2 -I.  -c -o kfunc.o kfunc.c\n",
            "gcc -Wall -g -O2 -I.  -c -o knetfile.o knetfile.c\n",
            "gcc -Wall -g -O2 -I.  -c -o kstring.o kstring.c\n",
            "gcc -Wall -g -O2 -I.  -c -o bcf_sr_sort.o bcf_sr_sort.c\n",
            "gcc -Wall -g -O2 -I.  -c -o bgzf.o bgzf.c\n",
            "gcc -Wall -g -O2 -I.  -c -o errmod.o errmod.c\n",
            "gcc -Wall -g -O2 -I.  -c -o faidx.o faidx.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile.o hfile.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile_net.o hfile_net.c\n",
            "echo '#define HTS_VERSION \"1.9\"' > version.h\n",
            "gcc -Wall -g -O2 -I.  -c -o hts.o hts.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hts_os.o hts_os.c\n",
            "gcc -Wall -g -O2 -I.  -c -o md5.o md5.c\n",
            "gcc -Wall -g -O2 -I.  -c -o multipart.o multipart.c\n",
            "gcc -Wall -g -O2 -I.  -c -o probaln.o probaln.c\n",
            "gcc -Wall -g -O2 -I.  -c -o realn.o realn.c\n",
            "gcc -Wall -g -O2 -I.  -c -o regidx.o regidx.c\n",
            "gcc -Wall -g -O2 -I.  -c -o sam.o sam.c\n",
            "gcc -Wall -g -O2 -I.  -c -o synced_bcf_reader.o synced_bcf_reader.c\n",
            "gcc -Wall -g -O2 -I.  -c -o vcf_sweep.o vcf_sweep.c\n",
            "gcc -Wall -g -O2 -I.  -c -o tbx.o tbx.c\n",
            "gcc -Wall -g -O2 -I.  -c -o textutils.o textutils.c\n",
            "gcc -Wall -g -O2 -I.  -c -o thread_pool.o thread_pool.c\n",
            "gcc -Wall -g -O2 -I.  -c -o vcf.o vcf.c\n",
            "gcc -Wall -g -O2 -I.  -c -o vcfutils.o vcfutils.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_codecs.o cram/cram_codecs.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_decode.o cram/cram_decode.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_encode.o cram/cram_encode.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_external.o cram/cram_external.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_index.o cram/cram_index.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_io.o cram/cram_io.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_samtools.o cram/cram_samtools.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/cram_stats.o cram/cram_stats.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/files.o cram/files.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/mFILE.o cram/mFILE.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/open_trace_file.o cram/open_trace_file.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/pooled_alloc.o cram/pooled_alloc.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/rANS_static.o cram/rANS_static.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/sam_header.o cram/sam_header.c\n",
            "gcc -Wall -g -O2 -I.  -c -o cram/string_alloc.o cram/string_alloc.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile_libcurl.o hfile_libcurl.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile_gcs.o hfile_gcs.c\n",
            "gcc -Wall -g -O2 -I.  -c -o hfile_s3.o hfile_s3.c\n",
            "ar -rc libhts.a kfunc.o knetfile.o kstring.o bcf_sr_sort.o bgzf.o errmod.o faidx.o hfile.o hfile_net.o hts.o hts_os.o md5.o multipart.o probaln.o realn.o regidx.o sam.o synced_bcf_reader.o vcf_sweep.o tbx.o textutils.o thread_pool.o vcf.o vcfutils.o cram/cram_codecs.o cram/cram_decode.o cram/cram_encode.o cram/cram_external.o cram/cram_index.o cram/cram_io.o cram/cram_samtools.o cram/cram_stats.o cram/files.o cram/mFILE.o cram/open_trace_file.o cram/pooled_alloc.o cram/rANS_static.o cram/sam_header.o cram/string_alloc.o  hfile_libcurl.o hfile_gcs.o hfile_s3.o\n",
            "ranlib libhts.a\n",
            "make[1]: Leaving directory '/content/samtools-1.9/htslib-1.9'\n",
            "gcc  -L./lz4  -o samtools bam_index.o bam_plcmd.o sam_view.o bam_cat.o bam_md.o bam_reheader.o bam_sort.o bedidx.o bam_rmdup.o bam_rmdupse.o bam_mate.o bam_stat.o bam_color.o bamtk.o bam2bcf.o bam2bcf_indel.o sample.o cut_target.o phase.o bam2depth.o padding.o bedcov.o bamshuf.o faidx.o dict.o stats.o stats_isize.o bam_flags.o bam_split.o bam_tview.o bam_tview_curses.o bam_tview_html.o bam_lpileup.o bam_quickcheck.o bam_addrprg.o bam_markdup.o tmp_file.o ./lz4/lz4.o libbam.a libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lncurses -lm -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/ace2sam.o misc/ace2sam.c\n",
            "gcc  -o misc/ace2sam misc/ace2sam.o -lz \n",
            "gcc -g -O2 -DMAQ_LONGREADS -I. -Ihtslib-1.9 -I./lz4  -c -o misc/maq2sam-long.o misc/maq2sam.c\n",
            "gcc  -o misc/maq2sam-long misc/maq2sam-long.o -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/maq2sam-short.o misc/maq2sam.c\n",
            "gcc  -o misc/maq2sam-short misc/maq2sam-short.o -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/md5fa.o misc/md5fa.c\n",
            "gcc  -L./lz4  -o misc/md5fa misc/md5fa.o htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/md5sum-lite.o misc/md5sum-lite.c\n",
            "gcc  -L./lz4  -o misc/md5sum-lite misc/md5sum-lite.o htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o misc/wgsim.o misc/wgsim.c\n",
            "gcc  -L./lz4  -o misc/wgsim misc/wgsim.o -lm htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz \n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/merge/test_bam_translate.o test/merge/test_bam_translate.c\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/test.o test/test.c\n",
            "gcc  -L./lz4  -o test/merge/test_bam_translate test/merge/test_bam_translate.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/merge/test_rtrans_build.o test/merge/test_rtrans_build.c\n",
            "gcc  -L./lz4  -o test/merge/test_rtrans_build test/merge/test_rtrans_build.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/merge/test_trans_tbl_init.o test/merge/test_trans_tbl_init.c\n",
            "gcc  -L./lz4  -o test/merge/test_trans_tbl_init test/merge/test_trans_tbl_init.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/split/test_count_rg.o test/split/test_count_rg.c\n",
            "gcc  -L./lz4  -o test/split/test_count_rg test/split/test_count_rg.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/split/test_expand_format_string.o test/split/test_expand_format_string.c\n",
            "gcc  -L./lz4  -o test/split/test_expand_format_string test/split/test_expand_format_string.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/split/test_filter_header_rg.o test/split/test_filter_header_rg.c\n",
            "gcc  -L./lz4  -o test/split/test_filter_header_rg test/split/test_filter_header_rg.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/split/test_parse_args.o test/split/test_parse_args.c\n",
            "gcc  -L./lz4  -o test/split/test_parse_args test/split/test_parse_args.o test/test.o libst.a htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "gcc -g -O2 -I. -Ihtslib-1.9 -I./lz4  -c -o test/vcf-miniview.o test/vcf-miniview.c\n",
            "gcc  -L./lz4  -o test/vcf-miniview test/vcf-miniview.o htslib-1.9/libhts.a -lpthread -lz -lm -lbz2 -llzma -lcurl -lcrypto -lz  -lpthread\n",
            "mkdir -p -m 755 /usr/local/bin /usr/local/bin /usr/local/share/man/man1\n",
            "install -p samtools /usr/local/bin\n",
            "install -p misc/ace2sam misc/maq2sam-long misc/maq2sam-short misc/md5fa misc/md5sum-lite misc/wgsim /usr/local/bin\n",
            "install -p misc/blast2sam.pl misc/bowtie2sam.pl misc/export2sam.pl misc/interpolate_sam.pl misc/novo2sam.pl misc/plot-bamstats misc/psl2sam.pl misc/sam2vcf.pl misc/samtools.pl misc/seq_cache_populate.pl misc/soap2sam.pl misc/varfilter.py misc/wgsim_eval.pl misc/zoom2sam.pl /usr/local/bin\n",
            "install -p -m 644 samtools.1 misc/wgsim.1 /usr/local/share/man/man1\n",
            "/content\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BEEYUWloHvzu",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 394
        },
        "outputId": "fb82e6d5-c8f7-4b49-8f7c-2d60a9addf3f"
      },
      "source": [
        "![[ -f ENCFF960ZGP.bed.gz ]] || wget https://www.encodeproject.org/files/ENCFF960ZGP/@@download/ENCFF960ZGP.bed.gz\n",
        "!ln -s ENCFF960ZGP.bed.gz peaks_with_signal.bed.gz"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-07-15 19:45:53--  https://www.encodeproject.org/files/ENCFF960ZGP/@@download/ENCFF960ZGP.bed.gz\n",
            "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
            "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
            "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
            "Location: https://download.encodeproject.org/https://encode-public.s3.amazonaws.com/2017/03/24/8be54b61-96de-4a54-a788-42f1f31c5370/ENCFF960ZGP.bed.gz?response-content-disposition=attachment%3B%20filename%3DENCFF960ZGP.bed.gz&Expires=1563349553&AWSAccessKeyId=ASIATGZNGCNXVP6QSJ4R&Signature=WPMHs13HsPZlJQoLwgMx9IOrGv0%3D&x-amz-security-token=AgoJb3JpZ2luX2VjELP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQCBU%2BCxnlAqyhwHMirCyU7rTednSqumAAQ2ELzzAdBWvgIhAJz%2B5G89qmVY5VDN7YQHWnOivuxQVMYwynQruWwkz3APKtoDCBwQABoMMjIwNzQ4NzE0ODYzIgzqtehKREjqrYtO3qAqtwNarSt0oxnG3FsyueNygF%2FerEyr9ejEL7Kr8kqWOxKgtei3%2F0b3UjeEu4wHNMe2DDiE0oaebZya8E89x%2FVyaFBmRP7ph7Q01yK%2FgD21N5ltb%2FfQOMAC3MrNGKRKs5JsDxsSt86cPjnsgo7SwnLiFVlr8GzB%2BjHk%2BfjPjsbO1hBorNAvnCjjhZLEZs431W1O8XVPEoI8DeMzrwsR8wTV0eTs6Ht74W4p5ONd14MIA%2BudryVOL7MYHIiDhBcU4sdQUdGsZQwzwKFTaHRpwJ3JjMzPAPRq%2B8IZW0juc8d5dgxtHT%2F0JR8xzstXwymvI4z2wl7g1G3D1U4DJBNA9v%2FGqG1Wo4jGrLOBaVRDuU2z90ZONs6hAC90%2FNp0aLEEiRqQEX4O9pDebO2rVCfTlJa6uieV8arsIdk%2FX0bmmJNpwgypz2miwKr2QvWliOkfRBRbIFPrhVoPp8SEywCYg5bTFnRTdaTNE%2BN1%2FHDhFPui4BtmUI6mFxsxVRJPMZly3%2BkollpsqZj7dM9GxobdZ4gcf608lCrh7RYsdMSQfiZffpcCTKh8Ph%2BNK%2BDMwvd3bazoXFAyc1yWddj1MNGVs%2BkFOrMBk7sWUQriz1yrgWBTq0L4YnQxDAGbNmc5ugPKtaRBXmPNR%2FiHI9aOklqqO%2BBs0hnpqpOcheDTAmLjvHt39xkwdJER%2BUJo76gppB2CI7Ly7y9lUAJLcAkvf8TG14B6dHS1HlFFLJ0lYyQA97LiG%2BA2Wa1Wm95KiEalHmxzziuzBRtQfXJBHKpFrJVSsWP1OUeJ%2FWWmzH4GkKq8Ld4PBub1Gbzs7sW%2BV9ctHW09mml1xsdNVBI%3D [following]\n",
            "--2019-07-15 19:45:53--  https://download.encodeproject.org/https://encode-public.s3.amazonaws.com/2017/03/24/8be54b61-96de-4a54-a788-42f1f31c5370/ENCFF960ZGP.bed.gz?response-content-disposition=attachment%3B%20filename%3DENCFF960ZGP.bed.gz&Expires=1563349553&AWSAccessKeyId=ASIATGZNGCNXVP6QSJ4R&Signature=WPMHs13HsPZlJQoLwgMx9IOrGv0%3D&x-amz-security-token=AgoJb3JpZ2luX2VjELP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQCBU%2BCxnlAqyhwHMirCyU7rTednSqumAAQ2ELzzAdBWvgIhAJz%2B5G89qmVY5VDN7YQHWnOivuxQVMYwynQruWwkz3APKtoDCBwQABoMMjIwNzQ4NzE0ODYzIgzqtehKREjqrYtO3qAqtwNarSt0oxnG3FsyueNygF%2FerEyr9ejEL7Kr8kqWOxKgtei3%2F0b3UjeEu4wHNMe2DDiE0oaebZya8E89x%2FVyaFBmRP7ph7Q01yK%2FgD21N5ltb%2FfQOMAC3MrNGKRKs5JsDxsSt86cPjnsgo7SwnLiFVlr8GzB%2BjHk%2BfjPjsbO1hBorNAvnCjjhZLEZs431W1O8XVPEoI8DeMzrwsR8wTV0eTs6Ht74W4p5ONd14MIA%2BudryVOL7MYHIiDhBcU4sdQUdGsZQwzwKFTaHRpwJ3JjMzPAPRq%2B8IZW0juc8d5dgxtHT%2F0JR8xzstXwymvI4z2wl7g1G3D1U4DJBNA9v%2FGqG1Wo4jGrLOBaVRDuU2z90ZONs6hAC90%2FNp0aLEEiRqQEX4O9pDebO2rVCfTlJa6uieV8arsIdk%2FX0bmmJNpwgypz2miwKr2QvWliOkfRBRbIFPrhVoPp8SEywCYg5bTFnRTdaTNE%2BN1%2FHDhFPui4BtmUI6mFxsxVRJPMZly3%2BkollpsqZj7dM9GxobdZ4gcf608lCrh7RYsdMSQfiZffpcCTKh8Ph%2BNK%2BDMwvd3bazoXFAyc1yWddj1MNGVs%2BkFOrMBk7sWUQriz1yrgWBTq0L4YnQxDAGbNmc5ugPKtaRBXmPNR%2FiHI9aOklqqO%2BBs0hnpqpOcheDTAmLjvHt39xkwdJER%2BUJo76gppB2CI7Ly7y9lUAJLcAkvf8TG14B6dHS1HlFFLJ0lYyQA97LiG%2BA2Wa1Wm95KiEalHmxzziuzBRtQfXJBHKpFrJVSsWP1OUeJ%2FWWmzH4GkKq8Ld4PBub1Gbzs7sW%2BV9ctHW09mml1xsdNVBI%3D\n",
            "Resolving download.encodeproject.org (download.encodeproject.org)... 34.211.244.144\n",
            "Connecting to download.encodeproject.org (download.encodeproject.org)|34.211.244.144|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 Moved Permanently\n",
            "Location: https://encode-public.s3.amazonaws.com/2017/03/24/8be54b61-96de-4a54-a788-42f1f31c5370/ENCFF960ZGP.bed.gz?response-content-disposition=attachment%3B%20filename%3DENCFF960ZGP.bed.gz&Expires=1563349553&AWSAccessKeyId=ASIATGZNGCNXVP6QSJ4R&Signature=WPMHs13HsPZlJQoLwgMx9IOrGv0%3D&x-amz-security-token=AgoJb3JpZ2luX2VjELP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQCBU%2BCxnlAqyhwHMirCyU7rTednSqumAAQ2ELzzAdBWvgIhAJz%2B5G89qmVY5VDN7YQHWnOivuxQVMYwynQruWwkz3APKtoDCBwQABoMMjIwNzQ4NzE0ODYzIgzqtehKREjqrYtO3qAqtwNarSt0oxnG3FsyueNygF%2FerEyr9ejEL7Kr8kqWOxKgtei3%2F0b3UjeEu4wHNMe2DDiE0oaebZya8E89x%2FVyaFBmRP7ph7Q01yK%2FgD21N5ltb%2FfQOMAC3MrNGKRKs5JsDxsSt86cPjnsgo7SwnLiFVlr8GzB%2BjHk%2BfjPjsbO1hBorNAvnCjjhZLEZs431W1O8XVPEoI8DeMzrwsR8wTV0eTs6Ht74W4p5ONd14MIA%2BudryVOL7MYHIiDhBcU4sdQUdGsZQwzwKFTaHRpwJ3JjMzPAPRq%2B8IZW0juc8d5dgxtHT%2F0JR8xzstXwymvI4z2wl7g1G3D1U4DJBNA9v%2FGqG1Wo4jGrLOBaVRDuU2z90ZONs6hAC90%2FNp0aLEEiRqQEX4O9pDebO2rVCfTlJa6uieV8arsIdk%2FX0bmmJNpwgypz2miwKr2QvWliOkfRBRbIFPrhVoPp8SEywCYg5bTFnRTdaTNE%2BN1%2FHDhFPui4BtmUI6mFxsxVRJPMZly3%2BkollpsqZj7dM9GxobdZ4gcf608lCrh7RYsdMSQfiZffpcCTKh8Ph%2BNK%2BDMwvd3bazoXFAyc1yWddj1MNGVs%2BkFOrMBk7sWUQriz1yrgWBTq0L4YnQxDAGbNmc5ugPKtaRBXmPNR%2FiHI9aOklqqO%2BBs0hnpqpOcheDTAmLjvHt39xkwdJER%2BUJo76gppB2CI7Ly7y9lUAJLcAkvf8TG14B6dHS1HlFFLJ0lYyQA97LiG%2BA2Wa1Wm95KiEalHmxzziuzBRtQfXJBHKpFrJVSsWP1OUeJ%2FWWmzH4GkKq8Ld4PBub1Gbzs7sW%2BV9ctHW09mml1xsdNVBI%3D [following]\n",
            "--2019-07-15 19:45:53--  https://encode-public.s3.amazonaws.com/2017/03/24/8be54b61-96de-4a54-a788-42f1f31c5370/ENCFF960ZGP.bed.gz?response-content-disposition=attachment%3B%20filename%3DENCFF960ZGP.bed.gz&Expires=1563349553&AWSAccessKeyId=ASIATGZNGCNXVP6QSJ4R&Signature=WPMHs13HsPZlJQoLwgMx9IOrGv0%3D&x-amz-security-token=AgoJb3JpZ2luX2VjELP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQCBU%2BCxnlAqyhwHMirCyU7rTednSqumAAQ2ELzzAdBWvgIhAJz%2B5G89qmVY5VDN7YQHWnOivuxQVMYwynQruWwkz3APKtoDCBwQABoMMjIwNzQ4NzE0ODYzIgzqtehKREjqrYtO3qAqtwNarSt0oxnG3FsyueNygF%2FerEyr9ejEL7Kr8kqWOxKgtei3%2F0b3UjeEu4wHNMe2DDiE0oaebZya8E89x%2FVyaFBmRP7ph7Q01yK%2FgD21N5ltb%2FfQOMAC3MrNGKRKs5JsDxsSt86cPjnsgo7SwnLiFVlr8GzB%2BjHk%2BfjPjsbO1hBorNAvnCjjhZLEZs431W1O8XVPEoI8DeMzrwsR8wTV0eTs6Ht74W4p5ONd14MIA%2BudryVOL7MYHIiDhBcU4sdQUdGsZQwzwKFTaHRpwJ3JjMzPAPRq%2B8IZW0juc8d5dgxtHT%2F0JR8xzstXwymvI4z2wl7g1G3D1U4DJBNA9v%2FGqG1Wo4jGrLOBaVRDuU2z90ZONs6hAC90%2FNp0aLEEiRqQEX4O9pDebO2rVCfTlJa6uieV8arsIdk%2FX0bmmJNpwgypz2miwKr2QvWliOkfRBRbIFPrhVoPp8SEywCYg5bTFnRTdaTNE%2BN1%2FHDhFPui4BtmUI6mFxsxVRJPMZly3%2BkollpsqZj7dM9GxobdZ4gcf608lCrh7RYsdMSQfiZffpcCTKh8Ph%2BNK%2BDMwvd3bazoXFAyc1yWddj1MNGVs%2BkFOrMBk7sWUQriz1yrgWBTq0L4YnQxDAGbNmc5ugPKtaRBXmPNR%2FiHI9aOklqqO%2BBs0hnpqpOcheDTAmLjvHt39xkwdJER%2BUJo76gppB2CI7Ly7y9lUAJLcAkvf8TG14B6dHS1HlFFLJ0lYyQA97LiG%2BA2Wa1Wm95KiEalHmxzziuzBRtQfXJBHKpFrJVSsWP1OUeJ%2FWWmzH4GkKq8Ld4PBub1Gbzs7sW%2BV9ctHW09mml1xsdNVBI%3D\n",
            "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.218.248.98\n",
            "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.218.248.98|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 705627 (689K) [binary/octet-stream]\n",
            "Saving to: ‘ENCFF960ZGP.bed.gz’\n",
            "\n",
            "ENCFF960ZGP.bed.gz  100%[===================>] 689.09K  2.68MB/s    in 0.3s    \n",
            "\n",
            "2019-07-15 19:45:54 (2.68 MB/s) - ‘ENCFF960ZGP.bed.gz’ saved [705627/705627]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sIMQ4EXcHxVU",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "#We want to prepare a bed file that has +/- 1kb around the summit, followed by\n",
        "# the signal strength\n",
        "! zcat peaks_with_signal.bed.gz | perl -lane 'print $F[0].\"\\t\".($F[1]+$F[9]).\"\\t\".($F[1]+$F[9]).\"\\t+\\t\".($F[6])' | egrep -w 'chr1|chr2|chr3|chr4|chr5|chr6|chr7|chr8|chr9|chr10|chr11|chr12|chr13|chr14|chr15|chr16|chr17|chr18|chr19|chr20|chr21|chr22|chrX|chrY' | gzip -c > summits_with_signal.bed.gz\n",
        "\n",
        "#We split into training/test/validation set by chromosome\n",
        "!zcat summits_with_signal.bed.gz | egrep -w 'chr1|chr8|chr21' | gzip -c > test_summits_with_signal.bed.gz\n",
        "!zcat summits_with_signal.bed.gz | egrep -w 'chr22' | gzip -c > valid_summits_with_signal.bed.gz\n",
        "!zcat summits_with_signal.bed.gz | egrep -w -v 'chr1|chr8|chr21|chr22' | gzip -c > train_summits_with_signal.bed.gz"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WcoKiHO3Hyuu",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "e30b524e-61bd-46a0-f719-1cb1a4e47324"
      },
      "source": [
        "![[ -e seqdataloader ]] && rm -rf seqdataloader\n",
        "!git clone https://github.com/kundajelab/seqdataloader.git\n",
        "%cd seqdataloader\n",
        "!pip uninstall seqdataloader\n",
        "!pip install .\n",
        "%cd .."
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'seqdataloader'...\n",
            "remote: Enumerating objects: 141, done.\u001b[K\n",
            "remote: Counting objects: 100% (141/141), done.\u001b[K\n",
            "remote: Compressing objects: 100% (93/93), done.\u001b[K\n",
            "remote: Total 649 (delta 90), reused 94 (delta 48), pack-reused 508\u001b[K\n",
            "Receiving objects: 100% (649/649), 3.80 MiB | 23.90 MiB/s, done.\n",
            "Resolving deltas: 100% (416/416), done.\n",
            "/content/seqdataloader\n",
            "\u001b[33mWARNING: Skipping seqdataloader as it is not installed.\u001b[0m\n",
            "Processing /content/seqdataloader\n",
            "Requirement already satisfied: numpy>=1.15 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (1.16.4)\n",
            "Requirement already satisfied: pandas>=0.23.4 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.24.2)\n",
            "Requirement already satisfied: cython>=0.27.3 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.29.12)\n",
            "Collecting deeptools>=3.0.1 (from seqdataloader==0.126)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/3d/16/3e1757b61db790c86d1d9cf189a80946785ee69a60648647e1a44bfe504f/deepTools-3.3.0.tar.gz (188kB)\n",
            "\u001b[K     |████████████████████████████████| 194kB 3.3MB/s \n",
            "\u001b[?25hCollecting pybedtools>=0.7 (from seqdataloader==0.126)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/ca/b6/af143d5247cfe331e32c96ca92056293140eb8ce788d37842f6dcea734b4/pybedtools-0.8.0.tar.gz (12.5MB)\n",
            "\u001b[K     |████████████████████████████████| 12.5MB 50.9MB/s \n",
            "\u001b[?25hCollecting pyBigWig>=0.3.2 (from seqdataloader==0.126)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/67/c8/741cca63e6a8688ef22a605bb06cc849fc57df419027013f2d1181f184f8/pyBigWig-0.3.16.tar.gz (65kB)\n",
            "\u001b[K     |████████████████████████████████| 71kB 30.6MB/s \n",
            "\u001b[?25hCollecting pyfaidx (from seqdataloader==0.126)\n",
            "  Downloading https://files.pythonhosted.org/packages/75/a5/7e2569527b3849ea28d79b4f70d7cf46a47d36459bc59e0efa4e10e8c8b2/pyfaidx-0.5.5.2.tar.gz\n",
            "Requirement already satisfied: python-dateutil>=2.5.0 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.23.4->seqdataloader==0.126) (2.5.3)\n",
            "Requirement already satisfied: pytz>=2011k in /usr/local/lib/python3.6/dist-packages (from pandas>=0.23.4->seqdataloader==0.126) (2018.9)\n",
            "Requirement already satisfied: scipy>=0.17.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (1.3.0)\n",
            "Requirement already satisfied: matplotlib>=2.1.2 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (3.0.3)\n",
            "Collecting pysam>=0.14.0 (from deeptools>=3.0.1->seqdataloader==0.126)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/f1/fc/d2be1a093bd8494ab63e3168aca36c2494753bbff190f3201ce2e7da9cda/pysam-0.15.2-cp36-cp36m-manylinux1_x86_64.whl (9.6MB)\n",
            "\u001b[K     |████████████████████████████████| 9.6MB 43.2MB/s \n",
            "\u001b[?25hCollecting numpydoc>=0.5 (from deeptools>=3.0.1->seqdataloader==0.126)\n",
            "  Downloading https://files.pythonhosted.org/packages/6a/f3/7cfe4c616e4b9fe05540256cc9c6661c052c8a4cec2915732793b36e1843/numpydoc-0.9.1.tar.gz\n",
            "Collecting py2bit>=0.2.0 (from deeptools>=3.0.1->seqdataloader==0.126)\n",
            "  Downloading https://files.pythonhosted.org/packages/53/bb/547a927bed736ead3dc909e1e552d57c9034bb9493eff80544c0cf6e4828/py2bit-0.3.0.tar.gz\n",
            "Requirement already satisfied: plotly>=2.0.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (3.6.1)\n",
            "Collecting deeptoolsintervals>=0.1.8 (from deeptools>=3.0.1->seqdataloader==0.126)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/5e/a6/fa0ec5fb29f0ab041ce69e6e837e04a831022bb5b1e2e8b524491d53b8a3/deeptoolsintervals-0.1.8.tar.gz (40kB)\n",
            "\u001b[K     |████████████████████████████████| 40kB 20.1MB/s \n",
            "\u001b[?25hRequirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from pybedtools>=0.7->seqdataloader==0.126) (1.12.0)\n",
            "Requirement already satisfied: setuptools>=0.7 in /usr/local/lib/python3.6/dist-packages (from pyfaidx->seqdataloader==0.126) (41.0.1)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (2.4.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (1.1.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (0.10.0)\n",
            "Requirement already satisfied: sphinx>=1.6.5 in /usr/local/lib/python3.6/dist-packages (from numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.8.5)\n",
            "Requirement already satisfied: Jinja2>=2.3 in /usr/local/lib/python3.6/dist-packages (from numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.10.1)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.21.0)\n",
            "Requirement already satisfied: nbformat>=4.2 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.4.0)\n",
            "Requirement already satisfied: retrying>=1.3.3 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (1.3.3)\n",
            "Requirement already satisfied: decorator>=4.0.6 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.4.0)\n",
            "Requirement already satisfied: Pygments>=2.0 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.1.3)\n",
            "Requirement already satisfied: sphinxcontrib-websupport in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.2)\n",
            "Requirement already satisfied: snowballstemmer>=1.1 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.9.0)\n",
            "Requirement already satisfied: docutils>=0.11 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (0.14)\n",
            "Requirement already satisfied: alabaster<0.8,>=0.7 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (0.7.12)\n",
            "Requirement already satisfied: babel!=2.0,>=1.3 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.7.0)\n",
            "Requirement already satisfied: imagesize in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (19.0)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.6/dist-packages (from Jinja2>=2.3->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.1)\n",
            "Requirement already satisfied: idna<2.9,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.8)\n",
            "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (3.0.4)\n",
            "Requirement already satisfied: urllib3<1.25,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (1.24.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2019.6.16)\n",
            "Requirement already satisfied: traitlets>=4.1 in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.3.2)\n",
            "Requirement already satisfied: ipython-genutils in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (0.2.0)\n",
            "Requirement already satisfied: jsonschema!=2.5.0,>=2.4 in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.6.0)\n",
            "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.5.0)\n",
            "Building wheels for collected packages: seqdataloader, deeptools, pybedtools, pyBigWig, pyfaidx, numpydoc, py2bit, deeptoolsintervals\n",
            "  Building wheel for seqdataloader (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /tmp/pip-ephem-wheel-cache-x1jmns2y/wheels/c2/db/13/112d41662f69fb8c7986c218293570cc1550fc21eed966e31b\n",
            "  Building wheel for deeptools (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/de/22/40/f5eb4ef7cb83c890596ce90260a478008adeca4e4138f64430\n",
            "  Building wheel for pybedtools (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/4b/62/5b/fcd4580cc7fd70075dc142673a677bed992b5217a7ce22b973\n",
            "  Building wheel for pyBigWig (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/d7/b9/a7/4ced1fb3a4e8816bbdc7bd3fee1d378d76e57e55ead6113976\n",
            "  Building wheel for pyfaidx (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/54/a2/b4/e242e58d23b2808e191b214067880faa46cd2341f363886e0b\n",
            "  Building wheel for numpydoc (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/51/30/d1/92a39ba40f21cb70e53f8af96eb98f002a781843c065406500\n",
            "  Building wheel for py2bit (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/df/66/b6/33fb9b65b31121127f1da60ca27948ecf8d4c59b0967298de8\n",
            "  Building wheel for deeptoolsintervals (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/64/c7/27/eac1d52700f324ae4c7e2455a6a7cca24d9bd3d2d68ffc1962\n",
            "Successfully built seqdataloader deeptools pybedtools pyBigWig pyfaidx numpydoc py2bit deeptoolsintervals\n",
            "Installing collected packages: pysam, numpydoc, pyBigWig, py2bit, deeptoolsintervals, deeptools, pybedtools, pyfaidx, seqdataloader\n",
            "Successfully installed deeptools-3.3.0 deeptoolsintervals-0.1.8 numpydoc-0.9.1 py2bit-0.3.0 pyBigWig-0.3.16 pybedtools-0.8.0 pyfaidx-0.5.5.2 pysam-0.15.2 seqdataloader-0.126\n",
            "/content\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xAW-jWd9H0Dl",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "outputId": "639b8a71-15f8-4e73-9620-c79ea177e36a"
      },
      "source": [
        "from seqdataloader.batchproducers import coordbased\n",
        "import gzip\n",
        "import numpy as np\n",
        "class SiameseAugmenter(coordbased.coordbatchtransformers.AbstractCoordBatchTransformer): \n",
        "  def __call__(self, coords): \n",
        "    return [x.get_revcomp() for x in coords]"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Using TensorFlow backend.\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aIZ6Jzy4H1ay",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "class ColsInBedFile(\n",
        "    coordbased.coordstovals.core.AbstractSingleNdarrayCoordsToVals):\n",
        "    def __init__(self, gzipped_bed_file, **kwargs):\n",
        "        super(ColsInBedFile, self).__init__(**kwargs)\n",
        "        self.gzipped_bed_file = gzipped_bed_file\n",
        "        coords_to_vals = {}\n",
        "        for row in gzip.open(gzipped_bed_file, 'rb'):\n",
        "            row = row.decode(\"utf-8\").rstrip()\n",
        "            split_row = row.split(\"\\t\")\n",
        "            chrom_start_end = split_row[0]+\":\"+split_row[1]+\"-\"+split_row[2]\n",
        "            vals = np.array([float(x) for x in split_row[4:]])\n",
        "            coords_to_vals[chrom_start_end] = vals\n",
        "        self.coords_to_vals = coords_to_vals\n",
        "        \n",
        "    def _get_ndarray(self, coors):\n",
        "        to_return = []\n",
        "        for coor in coors:\n",
        "            chrom_start_end = (coor.chrom+\":\"\n",
        "                               +str(coor.start)+\"-\"+str(coor.end))\n",
        "            to_return.append(self.coords_to_vals[chrom_start_end])\n",
        "        return np.array(to_return)\n",
        "    \n",
        "    \n",
        "inputs_coordstovals = coordbased.coordstovals.fasta.PyfaidxCoordsToVals(\n",
        "  genome_fasta_path=\"hg38.genome.fa\",\n",
        "  center_size_to_use=1000)\n",
        "\n",
        "targets_coordstovals = ColsInBedFile(\n",
        "       gzipped_bed_file=\"summits_with_signal.bed.gz\")\n",
        "            \n",
        "keras_train_batch_generator = coordbased.core.KerasBatchGenerator(\n",
        "    coordsbatch_producer=coordbased.coordbatchproducers.SimpleCoordsBatchProducer(\n",
        "      bed_file=\"train_summits_with_signal.bed.gz\",\n",
        "      batch_size=64,\n",
        "      shuffle_before_epoch=True,\n",
        "      seed=1234\n",
        "    ),\n",
        "    inputs_coordstovals=inputs_coordstovals,\n",
        "    targets_coordstovals=targets_coordstovals\n",
        ")\n",
        "\n",
        "keras_valid_batch_generator = coordbased.core.KerasBatchGenerator(\n",
        "    coordsbatch_producer = coordbased.coordbatchproducers.SimpleCoordsBatchProducer(\n",
        "        bed_file=\"valid_summits_with_signal.bed.gz\",\n",
        "        batch_size=64, \n",
        "        shuffle_before_epoch=True, \n",
        "        seed=1234\n",
        "    ),\n",
        "    inputs_coordstovals=inputs_coordstovals, \n",
        "    targets_coordstovals=targets_coordstovals\n",
        ")\n",
        "\n",
        "keras_test_batch_generator = coordbased.core.KerasBatchGenerator(\n",
        "    coordsbatch_producer = coordbased.coordbatchproducers.SimpleCoordsBatchProducer(\n",
        "        bed_file=\"test_summits_with_signal.bed.gz\", \n",
        "        batch_size = 64, \n",
        "        shuffle_before_epoch = True, \n",
        "        seed = 1234\n",
        "    ), \n",
        "    inputs_coordstovals = inputs_coordstovals, \n",
        "    targets_coordstovals = targets_coordstovals\n",
        ")"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LYkmD-MZH3Dq",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "y_test = np.array([val for batch in keras_test_batch_generator for val in batch[1]], dtype = 'float32') "
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "e3Mc6JdjH5Ec",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 952
        },
        "outputId": "a2c6f78c-5aa5-4da7-cba8-d23c6e78cd99"
      },
      "source": [
        "!pip install concise\n",
        "!pip install keras-genomics"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting concise\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/73/06/9b680a74cd7bc682c6da7da04704af89b4ea0eb635d917615e0c7401a686/concise-0.6.6.tar.gz (11.3MB)\n",
            "\u001b[K     |████████████████████████████████| 11.3MB 3.4MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.6/dist-packages (from concise) (1.16.4)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.6/dist-packages (from concise) (0.24.2)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.6/dist-packages (from concise) (1.3.0)\n",
            "Requirement already satisfied: scikit-learn>=0.18 in /usr/local/lib/python3.6/dist-packages (from concise) (0.21.2)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.6/dist-packages (from concise) (3.0.3)\n",
            "Requirement already satisfied: keras>=2.0.4 in /usr/local/lib/python3.6/dist-packages (from concise) (2.2.4)\n",
            "Requirement already satisfied: hyperopt in /usr/local/lib/python3.6/dist-packages (from concise) (0.1.2)\n",
            "Requirement already satisfied: descartes in /usr/local/lib/python3.6/dist-packages (from concise) (1.1.0)\n",
            "Requirement already satisfied: shapely in /usr/local/lib/python3.6/dist-packages (from concise) (1.6.4.post2)\n",
            "Collecting gtfparse>=1.0.7 (from concise)\n",
            "  Downloading https://files.pythonhosted.org/packages/41/5c/8bd2e9020051ccffc60c56ae70b32a3b649ddac1962e9aa641f93542440e/gtfparse-1.2.0.tar.gz\n",
            "Requirement already satisfied: pytz>=2011k in /usr/local/lib/python3.6/dist-packages (from pandas->concise) (2018.9)\n",
            "Requirement already satisfied: python-dateutil>=2.5.0 in /usr/local/lib/python3.6/dist-packages (from pandas->concise) (2.5.3)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.6/dist-packages (from scikit-learn>=0.18->concise) (0.13.2)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.6/dist-packages (from matplotlib->concise) (0.10.0)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib->concise) (2.4.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib->concise) (1.1.0)\n",
            "Requirement already satisfied: keras-preprocessing>=1.0.5 in /usr/local/lib/python3.6/dist-packages (from keras>=2.0.4->concise) (1.1.0)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.6/dist-packages (from keras>=2.0.4->concise) (3.13)\n",
            "Requirement already satisfied: six>=1.9.0 in /usr/local/lib/python3.6/dist-packages (from keras>=2.0.4->concise) (1.12.0)\n",
            "Requirement already satisfied: keras-applications>=1.0.6 in /usr/local/lib/python3.6/dist-packages (from keras>=2.0.4->concise) (1.0.8)\n",
            "Requirement already satisfied: h5py in /usr/local/lib/python3.6/dist-packages (from keras>=2.0.4->concise) (2.8.0)\n",
            "Requirement already satisfied: pymongo in /usr/local/lib/python3.6/dist-packages (from hyperopt->concise) (3.8.0)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.6/dist-packages (from hyperopt->concise) (2.3)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.6/dist-packages (from hyperopt->concise) (0.16.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.6/dist-packages (from hyperopt->concise) (4.28.1)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.6/dist-packages (from kiwisolver>=1.0.1->matplotlib->concise) (41.0.1)\n",
            "Requirement already satisfied: decorator>=4.3.0 in /usr/local/lib/python3.6/dist-packages (from networkx->hyperopt->concise) (4.4.0)\n",
            "Building wheels for collected packages: concise, gtfparse\n",
            "  Building wheel for concise (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/6e/9a/e1/2ca53dee91902d0d5a5137be3969e1535623928d08f6c47d58\n",
            "  Building wheel for gtfparse (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/c4/27/96/6ba6fe28cbb162c326823553e3e45ac502160d1340566360f8\n",
            "Successfully built concise gtfparse\n",
            "Installing collected packages: gtfparse, concise\n",
            "Successfully installed concise-0.6.6 gtfparse-1.2.0\n",
            "Collecting keras-genomics\n",
            "  Downloading https://files.pythonhosted.org/packages/3b/67/f8598275f0f1f210f0241177b841981192fff1a3d876c056afaafc8d1db7/keras_genomics-0.1.1.1.tar.gz\n",
            "Requirement already satisfied: keras in /usr/local/lib/python3.6/dist-packages (from keras-genomics) (2.2.4)\n",
            "Requirement already satisfied: numpy>=1.9.1 in /usr/local/lib/python3.6/dist-packages (from keras->keras-genomics) (1.16.4)\n",
            "Requirement already satisfied: h5py in /usr/local/lib/python3.6/dist-packages (from keras->keras-genomics) (2.8.0)\n",
            "Requirement already satisfied: keras-preprocessing>=1.0.5 in /usr/local/lib/python3.6/dist-packages (from keras->keras-genomics) (1.1.0)\n",
            "Requirement already satisfied: scipy>=0.14 in /usr/local/lib/python3.6/dist-packages (from keras->keras-genomics) (1.3.0)\n",
            "Requirement already satisfied: six>=1.9.0 in /usr/local/lib/python3.6/dist-packages (from keras->keras-genomics) (1.12.0)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.6/dist-packages (from keras->keras-genomics) (3.13)\n",
            "Requirement already satisfied: keras-applications>=1.0.6 in /usr/local/lib/python3.6/dist-packages (from keras->keras-genomics) (1.0.8)\n",
            "Building wheels for collected packages: keras-genomics\n",
            "  Building wheel for keras-genomics (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /root/.cache/pip/wheels/ad/07/53/12c9cab81be5fbb7f824df1cf6d23734f27ce7d52f0675691b\n",
            "Successfully built keras-genomics\n",
            "Installing collected packages: keras-genomics\n",
            "Successfully installed keras-genomics-0.1.1.1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "2Foizqd8H6en",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "1eb3bc82-fcac-4319-adb4-84453a439ab2"
      },
      "source": [
        "!pip install git+https://github.com/kundajelab/simdna.git@v0.4.3.1#egg=simdna\n",
        "!git clone https://github.com/kundajelab/revcomp_experiments.git\n",
        "%cd revcomp_experiments/\n",
        "!python setup.py install\n",
        "%cd .."
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting simdna from git+https://github.com/kundajelab/simdna.git@v0.4.3.1#egg=simdna\n",
            "  Cloning https://github.com/kundajelab/simdna.git (to revision v0.4.3.1) to /tmp/pip-install-9qcb2gux/simdna\n",
            "  Running command git clone -q https://github.com/kundajelab/simdna.git /tmp/pip-install-9qcb2gux/simdna\n",
            "  Running command git checkout -q ab8ff979761fa4bd307e2d75a87c1da258a9f743\n",
            "Requirement already satisfied: numpy>=1.9 in /usr/local/lib/python3.6/dist-packages (from simdna) (1.16.4)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.6/dist-packages (from simdna) (3.0.3)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.6/dist-packages (from simdna) (1.3.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib->simdna) (1.1.0)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib->simdna) (2.4.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.6/dist-packages (from matplotlib->simdna) (0.10.0)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib->simdna) (2.5.3)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.6/dist-packages (from kiwisolver>=1.0.1->matplotlib->simdna) (41.0.1)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from cycler>=0.10->matplotlib->simdna) (1.12.0)\n",
            "Building wheels for collected packages: simdna\n",
            "  Building wheel for simdna (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Stored in directory: /tmp/pip-ephem-wheel-cache-i1ladut9/wheels/d7/dd/b9/c0544822379aa265750e5c8e5e164435e70f9346cdac4516b4\n",
            "Successfully built simdna\n",
            "Installing collected packages: simdna\n",
            "Successfully installed simdna-0.4.3.1\n",
            "Cloning into 'revcomp_experiments'...\n",
            "remote: Enumerating objects: 3, done.\u001b[K\n",
            "remote: Counting objects: 100% (3/3), done.\u001b[K\n",
            "remote: Compressing objects: 100% (3/3), done.\u001b[K\n",
            "remote: Total 1381 (delta 0), reused 0 (delta 0), pack-reused 1378\u001b[K\n",
            "Receiving objects: 100% (1381/1381), 521.45 MiB | 38.03 MiB/s, done.\n",
            "Resolving deltas: 100% (308/308), done.\n",
            "Checking out files: 100% (1202/1202), done.\n",
            "/content/revcomp_experiments\n",
            "running install\n",
            "running bdist_egg\n",
            "running egg_info\n",
            "creating revcompexp.egg-info\n",
            "writing revcompexp.egg-info/PKG-INFO\n",
            "writing dependency_links to revcompexp.egg-info/dependency_links.txt\n",
            "writing requirements to revcompexp.egg-info/requires.txt\n",
            "writing top-level names to revcompexp.egg-info/top_level.txt\n",
            "writing manifest file 'revcompexp.egg-info/SOURCES.txt'\n",
            "writing manifest file 'revcompexp.egg-info/SOURCES.txt'\n",
            "installing library code to build/bdist.linux-x86_64/egg\n",
            "running install_lib\n",
            "warning: install_lib: 'build/lib' does not exist -- no Python modules to install\n",
            "\n",
            "creating build\n",
            "creating build/bdist.linux-x86_64\n",
            "creating build/bdist.linux-x86_64/egg\n",
            "creating build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "installing scripts to build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "running install_scripts\n",
            "running build_scripts\n",
            "creating build/scripts-3.6\n",
            "copying and adjusting scripts/motif_density_and_position_sim.py -> build/scripts-3.6\n",
            "changing mode of build/scripts-3.6/motif_density_and_position_sim.py from 644 to 755\n",
            "creating build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "copying build/scripts-3.6/motif_density_and_position_sim.py -> build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "changing mode of build/bdist.linux-x86_64/egg/EGG-INFO/scripts/motif_density_and_position_sim.py to 755\n",
            "copying revcompexp.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "copying revcompexp.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "copying revcompexp.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "copying revcompexp.egg-info/requires.txt -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "copying revcompexp.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "zip_safe flag not set; analyzing archive contents...\n",
            "creating dist\n",
            "creating 'dist/revcompexp-0.1.0.0-py3.6.egg' and adding 'build/bdist.linux-x86_64/egg' to it\n",
            "removing 'build/bdist.linux-x86_64/egg' (and everything under it)\n",
            "Processing revcompexp-0.1.0.0-py3.6.egg\n",
            "Copying revcompexp-0.1.0.0-py3.6.egg to /usr/local/lib/python3.6/dist-packages\n",
            "Adding revcompexp 0.1.0.0 to easy-install.pth file\n",
            "Installing motif_density_and_position_sim.py script to /usr/local/bin\n",
            "\n",
            "Installed /usr/local/lib/python3.6/dist-packages/revcompexp-0.1.0.0-py3.6.egg\n",
            "Processing dependencies for revcompexp==0.1.0.0\n",
            "Searching for scipy==1.3.0\n",
            "Best match: scipy 1.3.0\n",
            "Adding scipy 1.3.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for matplotlib==3.0.3\n",
            "Best match: matplotlib 3.0.3\n",
            "Adding matplotlib 3.0.3 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for numpy==1.16.4\n",
            "Best match: numpy 1.16.4\n",
            "Adding numpy 1.16.4 to easy-install.pth file\n",
            "Installing f2py script to /usr/local/bin\n",
            "Installing f2py3 script to /usr/local/bin\n",
            "Installing f2py3.6 script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for simdna==0.4.3.1\n",
            "Best match: simdna 0.4.3.1\n",
            "Adding simdna 0.4.3.1 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for kiwisolver==1.1.0\n",
            "Best match: kiwisolver 1.1.0\n",
            "Adding kiwisolver 1.1.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for cycler==0.10.0\n",
            "Best match: cycler 0.10.0\n",
            "Adding cycler 0.10.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for python-dateutil==2.5.3\n",
            "Best match: python-dateutil 2.5.3\n",
            "Adding python-dateutil 2.5.3 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for pyparsing==2.4.0\n",
            "Best match: pyparsing 2.4.0\n",
            "Adding pyparsing 2.4.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for setuptools==41.0.1\n",
            "Best match: setuptools 41.0.1\n",
            "Adding setuptools 41.0.1 to easy-install.pth file\n",
            "Installing easy_install script to /usr/local/bin\n",
            "Installing easy_install-3.6 script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Searching for six==1.12.0\n",
            "Best match: six 1.12.0\n",
            "Adding six 1.12.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.6/dist-packages\n",
            "Finished processing dependencies for revcompexp==0.1.0.0\n",
            "/content\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZYIP9FkcH8FN",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import keras \n",
        "import keras_genomics\n",
        "import numpy as np\n",
        "import keras.layers as k1\n",
        "\n",
        "from keras import backend as K \n",
        "from keras.layers.core import Dropout \n",
        "from keras.layers.core import Flatten\n",
        "from keras.layers import Input\n",
        "from keras.engine import Layer\n",
        "from keras.models import Sequential \n",
        "from keras.engine.base_layer import InputSpec\n",
        "from keras.models import Model\n",
        "from keras.models import load_model"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cXqiQSNvH9aW",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "kernel_size = 15\n",
        "filters= 15\n",
        "input_length = 1000\n",
        "\n",
        "from numpy.random import seed\n",
        "from tensorflow import set_random_seed\n",
        "from keras.callbacks import EarlyStopping, History, ModelCheckpoint\n",
        "\n",
        "seed_num = 1000\n",
        "seed(seed_num)\n",
        "set_random_seed(seed_num)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4RnRUYCGH-ss",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "class RevComp(Layer): \n",
        "    def __init__(self, **kwargs): \n",
        "      super(RevComp, self).__init__(**kwargs)\n",
        "\n",
        "    def build(self, input_shape):\n",
        "      super(RevComp, self).build(input_shape)\n",
        "\n",
        "    def call(self, inputs): \n",
        "      return inputs[:,::-1,::-1]\n",
        "      \n",
        "    def compute_output_shape(self, input_shape):\n",
        "      return input_shape"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "CAZLdKB6IADX",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 232
        },
        "outputId": "1f3c3360-8052-480f-bc6e-5761d7629771"
      },
      "source": [
        "s_model = Sequential([\n",
        "    k1.Conv1D(filters=filters, kernel_size=kernel_size,\n",
        "            input_shape=keras_train_batch_generator_reg[0][0].shape[1:], padding=\"same\"), \n",
        "#     k1.BatchNormalization(), \n",
        "    k1.core.Activation(\"relu\"),\n",
        "    k1.Conv1D(filters=filters, kernel_size=kernel_size,\n",
        "              padding=\"same\"), \n",
        "#     k1.BatchNormalization(), \n",
        "    k1.core.Activation(\"relu\"),\n",
        "    k1.Conv1D(filters=filters, kernel_size=kernel_size,\n",
        "              padding=\"same\"), \n",
        "#     k1.BatchNormalization(), \n",
        "    k1.core.Activation(\"relu\"),\n",
        "    k1.pooling.MaxPooling1D(pool_size=40,padding=\"same\",\n",
        "                                               strides=40), \n",
        "    Flatten(), \n",
        "    k1.Dense(units = 100, activation = \"relu\"),\n",
        "    k1.Dense(units = 1)\n",
        "], name = \"shared_layers\")\n",
        "\n",
        "s_model.summary()\n",
        "main_input = Input(shape=keras_train_batch_generator_reg[0][0].shape[1:])\n",
        "rev_input = Input(shape=keras_train_batch_generator_reg[0][0].shape[1:])\n",
        "\n",
        "rev_input = RevComp()(main_input)\n",
        "\n",
        "main_output = s_model(main_input)\n",
        "rev_output = s_model(rev_input)\n",
        "\n",
        "avg = k1.Average()([main_output, rev_output])\n",
        "siamese_model = Model(inputs = main_input, outputs = avg)\n",
        "\n",
        "merged = keras.layers.concatenate([main_output, rev_output])\n",
        "                                  \n",
        "siamese_model.compile(optimizer=\"adam\", loss=\"mean_squared_error\")\n",
        "early_stopping_callback = keras.callbacks.EarlyStopping(\n",
        "                              monitor='val_loss',\n",
        "                              patience= 60,\n",
        "                              restore_best_weights=True)\n",
        "siamese_model.fit_generator(generator= keras_train_batch_generator, \n",
        "                           epochs=200, callbacks=[early_stopping_callback],\n",
        "                           validation_data=keras_valid_batch_generator_reg)\n",
        "siamese_model.set_weights(early_stopping_callback.best_weights)                                                              "
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-14-2e888ac70498>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m s_model = Sequential([\n\u001b[1;32m      2\u001b[0m     k1.Conv1D(filters=filters, kernel_size=kernel_size,\n\u001b[0;32m----> 3\u001b[0;31m             input_shape=keras_train_batch_generator_reg[0][0].shape[1:], padding=\"same\"), \n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;31m#     k1.BatchNormalization(),\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mk1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mActivation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"relu\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'keras_train_batch_generator_reg' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HBa3vZq6IBkq",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "siamese_filename = ('siamese_%s.h5' % seed_num, str(seed_num))[0]\n",
        "siamese_model.save(siamese_filename)\n",
        "custom_objects = {\"RevComp\":RevComp}\n",
        "siamese_model_final = load_model(siamese_filename, custom_objects)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "80q4THhEIC_e",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "y_pred_siamese = siamese_model.predict_generator(keras_test_batch_generator)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "a5D-KTXiIEIa",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "from matplotlib import pyplot as plt\n",
        "from scipy.stats import spearmanr\n",
        "\n",
        "plt.scatter(y_test, y_pred_siamese, alpha = 0.1)\n",
        "plt.xlabel(\"True Labels\")\n",
        "plt.ylabel(\"Predicted Labels\")\n",
        "plt.show()\n",
        "print(spearmanr(y_test, y_pred_siamese))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Vv8zB_tTIFRl",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        ""
      ],
      "execution_count": 0,
      "outputs": []
    }
  ]
}