{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/kundajelab/revcomp_experiments/blob/master/RegressionExample.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "kyeOFCfq24Cb"
   },
   "outputs": [],
   "source": [
    "from __future__ import division, print_function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "CHHAxdHG3MmH"
   },
   "source": [
    "Run some stuff to set up the colab environment (won't need to do this on the cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "colab_type": "code",
    "id": "Ih2JfMl73Q4w",
    "outputId": "9e172220-42f5-48e3-ac87-9a01c5babb1c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading package lists... Done\n",
      "Building dependency tree       \n",
      "Reading state information... Done\n",
      "bedtools is already the newest version (2.26.0+dfsg-5).\n",
      "The following package was automatically installed and is no longer required:\n",
      "  libnvidia-common-410\n",
      "Use 'apt autoremove' to remove it.\n",
      "0 upgraded, 0 newly installed, 0 to remove and 4 not upgraded.\n",
      "/content\n",
      "/content/samtools-1.9\n",
      "checking for gcc... gcc\n",
      "checking whether the C compiler works... yes\n",
      "checking for C compiler default output file name... a.out\n",
      "checking for suffix of executables... \n",
      "checking whether we are cross compiling... no\n",
      "checking for suffix of object files... o\n",
      "checking whether we are using the GNU C compiler... yes\n",
      "checking whether gcc accepts -g... yes\n",
      "checking for gcc option to accept ISO C89... none needed\n",
      "checking for special C compiler options needed for large files... no\n",
      "checking for _FILE_OFFSET_BITS value needed for large files... no\n",
      "checking location of HTSlib source tree... htslib-1.9\n",
      "checking for NcursesW wide-character library... no\n",
      "checking for Ncurses library... yes\n",
      "checking for working ncurses/curses.h... no\n",
      "checking for working ncurses.h... yes\n",
      "checking for zlib.h... yes\n",
      "checking for inflate in -lz... yes\n",
      "checking for library containing regcomp... none required\n",
      "configure: creating ./config.status\n",
      "config.status: creating config.mk\n",
      "config.status: creating config.h\n",
      "config.status: config.h is unchanged\n",
      "=== configuring in htslib-1.9 (/content/samtools-1.9/htslib-1.9)\n",
      "configure: running /bin/bash ./configure --disable-option-checking '--prefix=/usr/local'  --cache-file=/dev/null --srcdir=.\n",
      "checking for gcc... gcc\n",
      "checking whether the C compiler works... yes\n",
      "checking for C compiler default output file name... a.out\n",
      "checking for suffix of executables... \n",
      "checking whether we are cross compiling... no\n",
      "checking for suffix of object files... o\n",
      "checking whether we are using the GNU C compiler... yes\n",
      "checking whether gcc accepts -g... yes\n",
      "checking for gcc option to accept ISO C89... none needed\n",
      "checking for ranlib... ranlib\n",
      "checking for grep that handles long lines and -e... /bin/grep\n",
      "checking for C compiler warning flags... -Wall\n",
      "checking for special C compiler options needed for large files... no\n",
      "checking for _FILE_OFFSET_BITS value needed for large files... no\n",
      "checking for _LARGEFILE_SOURCE value needed for large files... no\n",
      "checking shared library type for unknown-Linux... plain .so\n",
      "checking how to run the C preprocessor... gcc -E\n",
      "checking for egrep... /bin/grep -E\n",
      "checking for ANSI C header files... yes\n",
      "checking for sys/types.h... yes\n",
      "checking for sys/stat.h... yes\n",
      "checking for stdlib.h... yes\n",
      "checking for string.h... yes\n",
      "checking for memory.h... yes\n",
      "checking for strings.h... yes\n",
      "checking for inttypes.h... yes\n",
      "checking for stdint.h... yes\n",
      "checking for unistd.h... yes\n",
      "checking for stdlib.h... (cached) yes\n",
      "checking for unistd.h... (cached) yes\n",
      "checking for sys/param.h... yes\n",
      "checking for getpagesize... yes\n",
      "checking for working mmap... yes\n",
      "checking for gmtime_r... yes\n",
      "checking for fsync... yes\n",
      "checking for drand48... yes\n",
      "checking whether fdatasync is declared... yes\n",
      "checking for fdatasync... yes\n",
      "checking for library containing log... -lm\n",
      "checking for zlib.h... yes\n",
      "checking for inflate in -lz... yes\n",
      "checking for library containing recv... none required\n",
      "checking for bzlib.h... yes\n",
      "checking for BZ2_bzBuffToBuffCompress in -lbz2... yes\n",
      "checking for lzma.h... yes\n",
      "checking for lzma_easy_buffer_encode in -llzma... yes\n",
      "checking for libdeflate.h... no\n",
      "checking for libdeflate_deflate_compress in -ldeflate... no\n",
      "checking for curl_easy_pause in -lcurl... yes\n",
      "checking for CCHmac... no\n",
      "checking for library containing HMAC... -lcrypto\n",
      "checking whether PTHREAD_MUTEX_RECURSIVE is declared... yes\n",
      "configure: creating ./config.status\n",
      "config.status: creating config.mk\n",
      "config.status: creating htslib.pc.tmp\n",
      "config.status: creating config.h\n",
      "config.status: config.h is unchanged\n",
      "cd htslib-1.9 && make htslib_static.mk\n",
      "make[1]: Entering directory '/content/samtools-1.9/htslib-1.9'\n",
      "sed -n '/^static_libs=/s/[^=]*=/HTSLIB_static_LIBS = /p;/^static_ldflags=/s/[^=]*=/HTSLIB_static_LDFLAGS = /p' htslib.pc.tmp > htslib_static.mk\n",
      "make[1]: Leaving directory '/content/samtools-1.9/htslib-1.9'\n",
      "make: Nothing to be done for 'all'.\n",
      "mkdir -p -m 755 /usr/local/bin /usr/local/bin /usr/local/share/man/man1\n",
      "install -p samtools /usr/local/bin\n",
      "install -p misc/ace2sam misc/maq2sam-long misc/maq2sam-short misc/md5fa misc/md5sum-lite misc/wgsim /usr/local/bin\n",
      "install -p misc/blast2sam.pl misc/bowtie2sam.pl misc/export2sam.pl misc/interpolate_sam.pl misc/novo2sam.pl misc/plot-bamstats misc/psl2sam.pl misc/sam2vcf.pl misc/samtools.pl misc/seq_cache_populate.pl misc/soap2sam.pl misc/varfilter.py misc/wgsim_eval.pl misc/zoom2sam.pl /usr/local/bin\n",
      "install -p -m 644 samtools.1 misc/wgsim.1 /usr/local/share/man/man1\n",
      "/content\n"
     ]
    }
   ],
   "source": [
    "#download the human genome\n",
    "#Get hg38 fasta by download 2bit and then converting to fa\n",
    "#On the cluster, this file is stored in /mnt/data/annotations/...\n",
    "![[ -f hg38.2bit ]] || wget http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.2bit -O hg38.2bit  \n",
    "![[ -f twoBitToFa ]] || wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/twoBitToFa -O twoBitToFa\n",
    "!chmod a+x twoBitToFa\n",
    "![[ -f hg38.genome.fa ]] || ./twoBitToFa hg38.2bit hg38.genome.fa\n",
    "\n",
    "#download hg38 chromsizes file\n",
    "# This would also be in /mnt/data/annotations\n",
    "![[ -f hg38.chrom.sizes ]] || wget http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes -O hg38.chrom.sizes\n",
    "  \n",
    "#install bedtools\n",
    "!apt-get install bedtools\n",
    "\n",
    "#Install samtools\n",
    "%cd /content\n",
    "![[ -f samtools-1.9.tar.bz2 ]] || wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2\n",
    "!tar -xjf samtools-1.9.tar.bz2\n",
    "%cd samtools-1.9\n",
    "!./configure\n",
    "!make\n",
    "!make install\n",
    "%cd ..\n",
    "\n",
    "#make index for the human genome\n",
    "#These are already built for the files on the cluster\n",
    "![[ -e hg38.genome.fa.fai ]] || samtools faidx hg38.genome.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 33
    },
    "colab_type": "code",
    "id": "OBdPtj383VTw",
    "outputId": "89447287-abc5-4703-e87c-bdad361341ef"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ln: failed to create symbolic link 'peaks_with_signal.bed.gz': File exists\n"
     ]
    }
   ],
   "source": [
    "![[ -f ENCFF794GVQ.bed.gz ]] || wget https://www.encodeproject.org/files/ENCFF794GVQ/@@download/ENCFF794GVQ.bed.gz\n",
    "!ln -s ENCFF794GVQ.bed.gz peaks_with_signal.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "jPUF-_5X5d88"
   },
   "outputs": [],
   "source": [
    "#We want to prepare a bed file that has +/- 1kb around the summit, followed by\n",
    "# the signal strength\n",
    "! zcat peaks_with_signal.bed.gz | perl -lane 'print $F[0].\"\\t\".($F[1]+$F[9]).\"\\t\".($F[1]+$F[9]).\"\\t+\\t\".($F[6])' | gzip -c > summits_with_signal.bed.gz\n",
    "\n",
    "#We split into training/test/validation set by chromosome\n",
    "!zcat summits_with_signal.bed.gz | egrep -w 'chr1|chr8|chr21' | gzip -c > test_summits_with_signal.bed.gz\n",
    "!zcat summits_with_signal.bed.gz | egrep -w 'chr22' | gzip -c > valid_summits_with_signal.bed.gz\n",
    "!zcat summits_with_signal.bed.gz | egrep -w -v 'chr1|chr8|chr21|chr22' | gzip -c > train_summits_with_signal.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "colab_type": "code",
    "id": "OhMjoBPE50JY",
    "outputId": "6eba01ff-46dd-4393-a420-a78abc4261bb"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'seqdataloader'...\n",
      "remote: Enumerating objects: 116, done.\u001b[K\n",
      "remote: Counting objects: 100% (116/116), done.\u001b[K\n",
      "remote: Compressing objects: 100% (79/79), done.\u001b[K\n",
      "remote: Total 624 (delta 72), reused 77 (delta 37), pack-reused 508\u001b[K\n",
      "Receiving objects: 100% (624/624), 3.80 MiB | 43.24 MiB/s, done.\n",
      "Resolving deltas: 100% (398/398), done.\n",
      "/content/seqdataloader\n",
      "Uninstalling seqdataloader-0.126:\n",
      "  Would remove:\n",
      "    /usr/local/bin/genomewide_labels\n",
      "    /usr/local/lib/python3.6/dist-packages/seqdataloader-0.126.dist-info/*\n",
      "    /usr/local/lib/python3.6/dist-packages/seqdataloader/*\n",
      "Proceed (y/n)? y\n",
      "  Successfully uninstalled seqdataloader-0.126\n",
      "Processing /content/seqdataloader\n",
      "Requirement already satisfied: numpy>=1.15 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (1.16.4)\n",
      "Requirement already satisfied: pandas>=0.23.4 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.24.2)\n",
      "Requirement already satisfied: cython>=0.27.3 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.29.12)\n",
      "Requirement already satisfied: deeptools>=3.0.1 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (3.3.0)\n",
      "Requirement already satisfied: pybedtools>=0.7 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.8.0)\n",
      "Requirement already satisfied: pyBigWig>=0.3.2 in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.3.16)\n",
      "Requirement already satisfied: pyfaidx in /usr/local/lib/python3.6/dist-packages (from seqdataloader==0.126) (0.5.5.2)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.23.4->seqdataloader==0.126) (2.5.3)\n",
      "Requirement already satisfied: pytz>=2011k in /usr/local/lib/python3.6/dist-packages (from pandas>=0.23.4->seqdataloader==0.126) (2018.9)\n",
      "Requirement already satisfied: scipy>=0.17.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (1.3.0)\n",
      "Requirement already satisfied: plotly>=2.0.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (3.6.1)\n",
      "Requirement already satisfied: numpydoc>=0.5 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (0.9.1)\n",
      "Requirement already satisfied: py2bit>=0.2.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (0.3.0)\n",
      "Requirement already satisfied: deeptoolsintervals>=0.1.8 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (0.1.8)\n",
      "Requirement already satisfied: pysam>=0.14.0 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (0.15.2)\n",
      "Requirement already satisfied: matplotlib>=2.1.2 in /usr/local/lib/python3.6/dist-packages (from deeptools>=3.0.1->seqdataloader==0.126) (3.0.3)\n",
      "Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from pybedtools>=0.7->seqdataloader==0.126) (1.12.0)\n",
      "Requirement already satisfied: setuptools>=0.7 in /usr/local/lib/python3.6/dist-packages (from pyfaidx->seqdataloader==0.126) (41.0.1)\n",
      "Requirement already satisfied: requests in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.21.0)\n",
      "Requirement already satisfied: decorator>=4.0.6 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.4.0)\n",
      "Requirement already satisfied: nbformat>=4.2 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.4.0)\n",
      "Requirement already satisfied: retrying>=1.3.3 in /usr/local/lib/python3.6/dist-packages (from plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (1.3.3)\n",
      "Requirement already satisfied: sphinx>=1.6.5 in /usr/local/lib/python3.6/dist-packages (from numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.8.5)\n",
      "Requirement already satisfied: Jinja2>=2.3 in /usr/local/lib/python3.6/dist-packages (from numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.10.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (1.1.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (0.10.0)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib>=2.1.2->deeptools>=3.0.1->seqdataloader==0.126) (2.4.0)\n",
      "Requirement already satisfied: urllib3<1.25,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (1.24.3)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.8)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2019.6.16)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (3.0.4)\n",
      "Requirement already satisfied: ipython-genutils in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (0.2.0)\n",
      "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.5.0)\n",
      "Requirement already satisfied: jsonschema!=2.5.0,>=2.4 in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (2.6.0)\n",
      "Requirement already satisfied: traitlets>=4.1 in /usr/local/lib/python3.6/dist-packages (from nbformat>=4.2->plotly>=2.0.0->deeptools>=3.0.1->seqdataloader==0.126) (4.3.2)\n",
      "Requirement already satisfied: sphinxcontrib-websupport in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.2)\n",
      "Requirement already satisfied: snowballstemmer>=1.1 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.9.0)\n",
      "Requirement already satisfied: Pygments>=2.0 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.1.3)\n",
      "Requirement already satisfied: babel!=2.0,>=1.3 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (2.7.0)\n",
      "Requirement already satisfied: alabaster<0.8,>=0.7 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (0.7.12)\n",
      "Requirement already satisfied: packaging in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (19.0)\n",
      "Requirement already satisfied: docutils>=0.11 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (0.14)\n",
      "Requirement already satisfied: imagesize in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.0)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.6/dist-packages (from Jinja2>=2.3->numpydoc>=0.5->deeptools>=3.0.1->seqdataloader==0.126) (1.1.1)\n",
      "Building wheels for collected packages: seqdataloader\n",
      "  Building wheel for seqdataloader (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
      "  Stored in directory: /tmp/pip-ephem-wheel-cache-nkg7shj3/wheels/c2/db/13/112d41662f69fb8c7986c218293570cc1550fc21eed966e31b\n",
      "Successfully built seqdataloader\n",
      "Installing collected packages: seqdataloader\n",
      "Successfully installed seqdataloader-0.126\n",
      "/content\n"
     ]
    }
   ],
   "source": [
    "![[ -e seqdataloader ]] && rm -rf seqdataloader\n",
    "!git clone https://github.com/kundajelab/seqdataloader.git\n",
    "%cd seqdataloader\n",
    "!pip uninstall seqdataloader\n",
    "!pip install .\n",
    "%cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 180
    },
    "colab_type": "code",
    "id": "T8AWnEeP6Thm",
    "outputId": "1c26c060-c8c6-4206-90b3-e88d3e51a1d8"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr5\t40756289\t40756289\t3.19598\n",
      "chr9\t129488708\t129488708\t3.25621\n",
      "chr12\t67269154\t67269154\t3.33539\n",
      "chr9\t2281574\t2281574\t3.57597\n",
      "chr10\t118594719\t118594719\t3.60814\n",
      "chr6\t37354217\t37354217\t3.66291\n",
      "chr12\t121800504\t121800504\t3.78161\n",
      "chr20\t62362217\t62362217\t3.86367\n",
      "chr1\t42963108\t42963108\t4.00591\n",
      "chr10\t35127150\t35127150\t4.05166\n"
     ]
    }
   ],
   "source": [
    "!zcat summits_with_signal.bed.gz | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "FbHhcBySQ4ZF"
   },
   "outputs": [],
   "source": [
    "from seqdataloader.batchproducers import coordbased\n",
    "import gzip\n",
    "import numpy as np\n",
    "\n",
    "class ColsInBedFile(\n",
    "    coordbased.coordstovals.core.AbstractSingleNdarrayCoordsToVals):\n",
    "    def __init__(self, gzipped_bed_file, **kwargs):\n",
    "        super(ColsInBedFile, self).__init__(**kwargs)\n",
    "        self.gzipped_bed_file = gzipped_bed_file\n",
    "        coords_to_vals = {}\n",
    "        for row in gzip.open(gzipped_bed_file, 'rb'):\n",
    "            row = row.decode(\"utf-8\").rstrip()\n",
    "            split_row = row.split(\"\\t\")\n",
    "            chrom_start_end = split_row[0]+\":\"+split_row[1]+\"-\"+split_row[2]\n",
    "            vals = np.array([float(x) for x in split_row[4:]])\n",
    "            coords_to_vals[chrom_start_end] = vals\n",
    "        self.coords_to_vals = coords_to_vals\n",
    "        \n",
    "    def _get_ndarray(self, coors):\n",
    "        to_return = []\n",
    "        for coor in coors:\n",
    "            chrom_start_end = (coor.chrom+\":\"\n",
    "                               +str(coor.start)+\"-\"+str(coor.end))\n",
    "            to_return.append(self.coords_to_vals[chrom_start_end])\n",
    "        return np.array(to_return)\n",
    "    \n",
    "    \n",
    "inputs_coordstovals = coordbased.coordstovals.fasta.PyfaidxCoordsToVals(\n",
    "  genome_fasta_path=\"hg38.genome.fa\",\n",
    "  center_size_to_use=1000)\n",
    "\n",
    "targets_coordstovals = ColsInBedFile(\n",
    "       gzipped_bed_file=\"summits_with_signal.bed.gz\")\n",
    "            \n",
    "keras_train_batch_generator = coordbased.core.KerasBatchGenerator(\n",
    "    coordsbatch_producer=coordbased.coordbatchproducers.SimpleCoordsBatchProducer(\n",
    "      bed_file=\"train_summits_with_signal.bed.gz\",\n",
    "      #coord_batch_transformer=coordbased.coordbatchtransformers.ReverseComplementAugmenter(),\n",
    "      batch_size=64,\n",
    "      shuffle_before_epoch=True,\n",
    "      seed=1234\n",
    "    ),\n",
    "    inputs_coordstovals=inputs_coordstovals,\n",
    "    targets_coordstovals=targets_coordstovals\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "colab_type": "code",
    "id": "Jz_sP2c5S-Ac",
    "outputId": "29de419b-acb1-4209-a194-f5e83aae35e5"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[[0, 1, 0, 0],\n",
       "         [0, 1, 0, 0],\n",
       "         [0, 1, 0, 0],\n",
       "         ...,\n",
       "         [0, 0, 1, 0],\n",
       "         [0, 0, 1, 0],\n",
       "         [1, 0, 0, 0]],\n",
       " \n",
       "        [[0, 0, 1, 0],\n",
       "         [0, 0, 0, 1],\n",
       "         [0, 0, 0, 1],\n",
       "         ...,\n",
       "         [0, 0, 0, 1],\n",
       "         [1, 0, 0, 0],\n",
       "         [0, 0, 1, 0]],\n",
       " \n",
       "        [[0, 1, 0, 0],\n",
       "         [0, 0, 1, 0],\n",
       "         [0, 1, 0, 0],\n",
       "         ...,\n",
       "         [1, 0, 0, 0],\n",
       "         [1, 0, 0, 0],\n",
       "         [1, 0, 0, 0]],\n",
       " \n",
       "        ...,\n",
       " \n",
       "        [[0, 0, 1, 0],\n",
       "         [1, 0, 0, 0],\n",
       "         [0, 0, 0, 1],\n",
       "         ...,\n",
       "         [0, 0, 1, 0],\n",
       "         [0, 0, 0, 1],\n",
       "         [0, 0, 1, 0]],\n",
       " \n",
       "        [[0, 0, 1, 0],\n",
       "         [0, 0, 1, 0],\n",
       "         [1, 0, 0, 0],\n",
       "         ...,\n",
       "         [0, 1, 0, 0],\n",
       "         [0, 1, 0, 0],\n",
       "         [0, 1, 0, 0]],\n",
       " \n",
       "        [[1, 0, 0, 0],\n",
       "         [0, 0, 0, 1],\n",
       "         [0, 0, 1, 0],\n",
       "         ...,\n",
       "         [0, 0, 0, 1],\n",
       "         [0, 0, 0, 1],\n",
       "         [0, 0, 0, 1]]]), array([[ 20.37783],\n",
       "        [ 48.88749],\n",
       "        [ 16.15834],\n",
       "        [ 41.14349],\n",
       "        [ 14.44541],\n",
       "        [ 16.42809],\n",
       "        [ 57.11373],\n",
       "        [ 66.70588],\n",
       "        [ 35.39692],\n",
       "        [ 33.25365],\n",
       "        [ 30.04847],\n",
       "        [ 18.83769],\n",
       "        [ 17.97874],\n",
       "        [ 21.88407],\n",
       "        [187.76758],\n",
       "        [ 15.2367 ],\n",
       "        [ 27.16318],\n",
       "        [ 37.87462],\n",
       "        [ 20.17201],\n",
       "        [ 26.86819],\n",
       "        [ 50.78424],\n",
       "        [ 19.85736],\n",
       "        [ 20.47014],\n",
       "        [ 17.42589],\n",
       "        [ 13.05479],\n",
       "        [ 23.46802],\n",
       "        [ 34.54703],\n",
       "        [ 19.4842 ],\n",
       "        [ 20.15646],\n",
       "        [ 16.1366 ],\n",
       "        [  3.78161],\n",
       "        [ 35.50247],\n",
       "        [ 44.4304 ],\n",
       "        [ 24.47947],\n",
       "        [ 22.04081],\n",
       "        [ 60.83001],\n",
       "        [ 20.68594],\n",
       "        [ 17.0564 ],\n",
       "        [ 20.40688],\n",
       "        [ 16.23506],\n",
       "        [ 18.40571],\n",
       "        [121.74944],\n",
       "        [ 15.45649],\n",
       "        [ 18.22361],\n",
       "        [  4.50975],\n",
       "        [ 34.36113],\n",
       "        [ 18.88764],\n",
       "        [ 17.58233],\n",
       "        [ 32.10156],\n",
       "        [ 15.91376],\n",
       "        [ 48.69843],\n",
       "        [ 20.40713],\n",
       "        [ 26.59024],\n",
       "        [ 27.85806],\n",
       "        [ 32.71879],\n",
       "        [ 32.57038],\n",
       "        [ 56.1416 ],\n",
       "        [ 10.82494],\n",
       "        [ 96.67858],\n",
       "        [ 18.42398],\n",
       "        [101.09937],\n",
       "        [ 16.78172],\n",
       "        [ 34.92996],\n",
       "        [ 33.92709],\n",
       "        [ 20.37783],\n",
       "        [ 48.88749],\n",
       "        [ 16.15834],\n",
       "        [ 41.14349],\n",
       "        [ 14.44541],\n",
       "        [ 16.42809],\n",
       "        [ 57.11373],\n",
       "        [ 66.70588],\n",
       "        [ 35.39692],\n",
       "        [ 33.25365],\n",
       "        [ 30.04847],\n",
       "        [ 18.83769],\n",
       "        [ 17.97874],\n",
       "        [ 21.88407],\n",
       "        [187.76758],\n",
       "        [ 15.2367 ],\n",
       "        [ 27.16318],\n",
       "        [ 37.87462],\n",
       "        [ 20.17201],\n",
       "        [ 26.86819],\n",
       "        [ 50.78424],\n",
       "        [ 19.85736],\n",
       "        [ 20.47014],\n",
       "        [ 17.42589],\n",
       "        [ 13.05479],\n",
       "        [ 23.46802],\n",
       "        [ 34.54703],\n",
       "        [ 19.4842 ],\n",
       "        [ 20.15646],\n",
       "        [ 16.1366 ],\n",
       "        [  3.78161],\n",
       "        [ 35.50247],\n",
       "        [ 44.4304 ],\n",
       "        [ 24.47947],\n",
       "        [ 22.04081],\n",
       "        [ 60.83001],\n",
       "        [ 20.68594],\n",
       "        [ 17.0564 ],\n",
       "        [ 20.40688],\n",
       "        [ 16.23506],\n",
       "        [ 18.40571],\n",
       "        [121.74944],\n",
       "        [ 15.45649],\n",
       "        [ 18.22361],\n",
       "        [  4.50975],\n",
       "        [ 34.36113],\n",
       "        [ 18.88764],\n",
       "        [ 17.58233],\n",
       "        [ 32.10156],\n",
       "        [ 15.91376],\n",
       "        [ 48.69843],\n",
       "        [ 20.40713],\n",
       "        [ 26.59024],\n",
       "        [ 27.85806],\n",
       "        [ 32.71879],\n",
       "        [ 32.57038],\n",
       "        [ 56.1416 ],\n",
       "        [ 10.82494],\n",
       "        [ 96.67858],\n",
       "        [ 18.42398],\n",
       "        [101.09937],\n",
       "        [ 16.78172],\n",
       "        [ 34.92996],\n",
       "        [ 33.92709]]))"
      ]
     },
     "execution_count": 35,
     "metadata": {
      "tags": []
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#define your model\n",
    "model.fit_generator(..)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "1RlWjtXKU6r7"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "include_colab_link": true,
   "name": "RegressionExample.ipynb",
   "provenance": [],
   "version": "0.3.2"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
